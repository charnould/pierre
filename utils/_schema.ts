import { z } from 'zod'

/**
 * Represents a User schema definition using Zod.
 *
 * This schema validates the structure of a user object with the following properties:
 *
 * - `email`: A string that is trimmed and converted to lowercase.
 * - `role`: An enumerated string that can be one of 'administrator', 'contributor', or 'collaborator'.
 * - `config`: A string that is trimmed.
 * - `password_hash`: A string representing the hashed password of the user.
 */
export const User = z.object({
  email: z.string().trim().toLowerCase(),
  role: z.enum(['administrator', 'contributor', 'collaborator']).catch('collaborator'),
  config: z.string().trim(),
  password_hash: z.string()
})

/**
 * Represents a parsed user schema that extends the base `User` schema
 * by adding a `config` property. The `config` property is an array
 * of strings, allowing for additional configuration options.
 *
 * This schema can be used to validate and type-check user data
 * with additional configuration details.
 */
export const Parsed_User = User.extend({ config: z.array(z.string()) })

//
// Incoming SMS parsing schema
export const SMS = z
  .object({
    role: z.string(),
    config: z.string(),
    conv_id: z.string(),
    phone: z.string().nullable(),
    to: z.string(),
    content: z.string().trim()
  })
  .or(z.null())

//
// `./assets/Config` schema
export const Config = z
  .object({
    id: z.string(),
    display: z.string(),
    show: z.array(z.string()),

    custom_data: z.object({ format: z.function() }).or(z.object({})),

    api: z
      .array(
        z.object({
          key: z.enum(['WEBHOOK_KEY_1', 'WEBHOOK_KEY_2', 'WEBHOOK_KEY_3']),
          url: z.string(), // URL check is made elsewhere
          format: z.function().args(
            z.object({
              custom_data: z.array(z.string()),
              content: z.string(),
              role: z.string()
            })
          )
        })
      )
      .default([]),

    models: z
      .object({
        augment_with: z.string(),
        rerank_with: z.string(),
        answer_with: z.string()
      })
      .default({
        augment_with: "openai('gpt-4o-mini-2024-07-18')",
        rerank_with: "openai('gpt-4o-mini-2024-07-18')",
        answer_with: "openai('gpt-4o-mini-2024-07-18')"
      }),
    phone: z.string().nullable(),
    protected: z.boolean(),
    knowledge: z.object({
      community: z.boolean(),
      proprietary: z.object({
        public: z.boolean(),
        private: z.boolean()
      })
    }),
    audience: z.string(),
    persona: z.string(),
    greeting: z.array(z.string()),
    examples: z.array(z.string()),
    disclaimer: z.string().nullable()
  })
  .strict()

//
// Reflects datastore database schema
export const Reply = z.object({
  // Globals
  conv_id: z.string(),
  config: z.string().or(Config),
  role: z.enum(['assistant', 'user', 'system']).default('user'),
  timestamp: z.string().datetime({ offset: true }).nullish().default(null),
  content: z.string(),
  // Metadata
  metadata: z
    .object({
      // User
      user: z.string().trim().toLowerCase().nullish().default(null),
      // Topics
      topics: z.string().trim().toLowerCase().nullish().default(null),
      // Evaluation + satisfaction
      evaluation: z
        .object({
          // CSAT
          customer: z
            .object({
              score: z.coerce.number().nullish().default(null).catch(null),
              comment: z.string().nullish().default(null)
            })
            .default({}),
          // Social housing organization satisfaction
          organization: z
            .object({
              score: z.coerce.number().nullish().default(null).catch(null),
              comment: z.string().nullish().default(null)
            })
            .default({}),
          // AI generated customer satisfaction (CSAT)
          ai: z
            .object({
              score: z.coerce.number().nullish().default(null).catch(null),
              comment: z.string().nullish().default(null)
            })
            .default({})
        })
        .default({}),
      // LLM usage
      tokens: z
        .object({
          // cache : z.number().nullish().default(null),
          prompt: z.number().nullish().default(null),
          completion: z.number().nullish().default(null),
          total: z.number().nullish().default(null)
        })
        .default({})
    })
    .default({})
})

//
// Structured JSON LLM must output for each request
export const Augmented_Query = z.object({
  lang: z.string(),
  contains_profanity: z.boolean(),
  standalone_questions: z.array(z.string()),
  bm25_keywords: z.array(z.string()),
  search_queries: z.array(z.string())
})

//
// AI Context
export const AIContext = Reply.extend({ query: Augmented_Query.nullable().default(null) })
  .merge(
    z.object({
      chunks: z
        .object({
          community: z.array(z.string()).nullish().default([]),
          private: z.array(z.string()).nullish().default([]),
          public: z.array(z.string()).nullish().default([])
        })
        .default({
          community: [],
          private: [],
          public: []
        }),
      custom_data: z.object({ raw: z.array(z.string()), transformed: z.string().default('') }),
      conversation: z
        .array(
          z.object({
            role: z.enum(['assistant', 'user', 'system']),
            content: z.string()
          })
        )
        .default([])
    })
  )
  .refine(async (c) => {
    // Change config name for config content
    if (typeof c.config === 'string') {
      c.config = (await import(`../assets/${c.config}/config`)).default
    }

    // Format data to be the one wanted by API
    if (typeof c.config !== 'string' && 'format' in c.config.custom_data) {
      if (
        Array.isArray(c.custom_data.raw) &&
        c.custom_data.raw.length === 1 &&
        c.custom_data.raw[0] === ''
      ) {
        c.custom_data.transformed = ''
      } else {
        c.custom_data.transformed = c.config.custom_data.format(c.custom_data.raw) as string
      }
    }

    return true
  })

//
//
export type SMS = z.infer<typeof SMS>
export type User = z.infer<typeof User>
export type Parsed_User = z.infer<typeof Parsed_User>
export type Reply = z.infer<typeof Reply>
export type Config = z.infer<typeof Config>
export type AIContext = z.infer<typeof AIContext>
export type Augmented_Query = z.infer<typeof Augmented_Query>
