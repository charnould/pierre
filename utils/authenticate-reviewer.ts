import type { Context, Next } from 'hono'
import { getSignedCookie } from 'hono/cookie'

export const authenticate = async (c: Context, next: Next) => {
  const cookie = await getSignedCookie(c, Bun.env.AUTH_SECRET as string, 'pierre')
  if (cookie === undefined) c.set('is_auth', false)
  else c.set('is_auth', true)
  await next()
}
