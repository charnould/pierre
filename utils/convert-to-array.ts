/**
 * Converts a given text string into an array of strings by performing several transformations:
 * - Removes any content wrapped in `<think>` tags.
 * - Removes escaped double quotes (`\"`) and regular double quotes (`"`).
 * - Splits the string by the pipe character (`|`).
 * - Trims whitespace and removes newline characters from each resulting line.
 * - Filters out any empty strings from the result.
 *
 * @note This function is tested in tests/utils/convert-to-array.test.ts
 * @param text - The input string to be converted.
 * @returns An array of non-empty, cleaned strings.
 */
export const convert_to_array = (text: string): string[] =>
  text
    .replace(/<think>\s*(.*?)\s*<\/think>/, ' ')
    .replace(/\\"/g, '')
    .replace(/"/g, '')
    .split('|')
    .map((line) => line.replace(/\n/g, '').trim())
    .filter((line) => line.length > 0)
