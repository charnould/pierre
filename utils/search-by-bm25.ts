import { string } from 'zod'
import type { AIContext } from './_schema'
import { db } from './database'
import type { Db_Name } from './database'
import { Vector_Search_Result } from './search-by-vectors'
import { stem } from './stem-text'

// TODO: implement `carry` stemmer
// TODO: waht about `Fine-Tune BM25 Hyperparameters` (chunk length is irrelevant)
export const bm25_search = (keyword: string, context: AIContext) => {
  try {
    console.log(keyword)
    // Retrieve knowledge access permissions for the current context
    // (e.g., default, team, etc.) + Set a default for TS happiness
    let k = { community: true, proprietary: { public: false, private: false } }
    if (typeof context.config !== 'string') k = context.config.knowledge

    // Initialize the bm25 search results (format
    // must be the same as Vector_Search_Result)
    const r: Vector_Search_Result = { community: [], private: [], public: [] }

    // Perform searches based on knowledge access
    if (k.community) r.community = query_db('community', keyword)
    if (k.proprietary.public) r.public = query_db('proprietary.public', keyword)
    if (k.proprietary.private) r.private = query_db('proprietary.private', keyword)

    // Parse results and return them to ensure
    // TypeScript types align with the schema
    return Vector_Search_Result.parse(r)
  } catch (e) {
    console.error(e)
    return
  }
}

//
//
//
//
// Helpers + Typescrit type
//
//
//
//

//
// Query the database with the keyword to retrieve the matching chunks
export const query_db = (db_name: Db_Name, keyword: string) => {
  const stemmed_keywords = stem(keyword)

  if (stemmed_keywords !== '') {
    // The rank column is the negative bm25 score
    // https://www.sqlite.org/fts5.html#:~:text=The%20%22%2D1%22%20term,numerically%20lower%20scores.
    // https://alexgarcia.xyz/blog/2024/sqlite-vec-hybrid-search/index.html
    const data = db(db_name)
      .prepare(
        `
        SELECT
          c.chunk_hash,
          c.chunk_text,
          s.rank AS distance
        FROM stems s 
        JOIN chunks c ON s.rowid = c.rowid
        WHERE s.chunk_stem MATCH ?
        ORDER BY distance
        LIMIT 15;
      `
      )
      .all(stemmed_keywords) as {
      chunk_hash: string
      chunk_text: string
      distance: number
    }[]

    return data
  }

  return []
}
