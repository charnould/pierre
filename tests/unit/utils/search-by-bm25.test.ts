import { expect, test } from 'bun:test'
import { query_db } from '../../../utils/search-by-bm25'

test('should bm25 search works', async () => {
  const context = (await import('../../../tests/unit/utils/__mocks__/ai_context.json')).default
  context.current_context = 'default'

  // "problème de voisinage", "règlement de conflit",
  // "conseils voisinage", "médiation", "relations de voisinage"

  const r = query_db('community', 'Valérie Létard, née le 13 octobre 1962')

  expect(r).toMatchSnapshot()
})
