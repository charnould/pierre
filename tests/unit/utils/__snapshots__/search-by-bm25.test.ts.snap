// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`should search using bm25 and find a result 1`] = `
[
  {
    "chunk_hash": "8q8lwccyqVt6SgLEaKdk",
    "chunk_text": 
"# Valérie Létard

Valérie Létard, née le 13 octobre 1962 à Orchies (Nord), est une femme politique française.

Proche de Jean-Louis Borloo, elle adhère à plusieurs partis centristes au cours de son parcours politique (UDF, PSD, FD, NC, UDI).

Au niveau local, elle est présidente de la communauté d'agglomération de Valenciennes (2008-2016), vice-présidente du conseil régional des Hauts-de-France (2016-2017) et conseillère départementale du Nord (depuis 2021).

Élue plusieurs fois sénatrice et députée depuis 2001, elle exerce la fonction de vice-présidente du Sénat de 2017 à 2023.

Entre 2007 et 2010, sous la présidence de Nicolas Sarkozy, elle est secrétaire d'État, successivement chargée de la Solidarité puis des Technologies vertes et des Négociations sur le climat.

Dans le gouvernement Michel Barnier, formé en 2024, elle occupe la fonction de ministre du Logement et de la Rénovation urbaine. Elle ne conserve que le portefeuille du Logement dans le gouvernement François Bayrou qui succède au précédent.

## Biographie

### Enfance et études

Fille de l'ancien député européen, président du Valenciennes Football Club, Francis Decourrière, Valérie Létard est assistante sociale de formation. Elle est également titulaire d'un DESS de développement local et économie solidaire.

### Parcours politique

Adhérente du Parti social-démocrate, Valérie Létard participe, au début de son activité professionnelle\\[Quand ?\\], à la mise en application concrète de nouveaux dispositifs d'action sociale à la mairie de Saint-Amand-les-Eaux, dans le Nord. De ses débuts, elle acquiert la mise en œuvre de projets et du travail en équipe. Du travail social, elle évolue vers le développement social des quartiers et la politique de la ville. Directrice du contrat de ville à Valenciennes, Jean-Louis Borloo lui propose de rejoindre l'équipe municipale comme adjointe au maire responsable de la politique de la ville.

#### Conseillère régionale et sénatrice du Nord

En 1998, elle est élue conseillère régionale et devient vice-présidente de la commission habitat-politique de la ville au conseil régional du Nord-Pas-de-Calais. En septembre 2001, lors des élections sénatoriales, elle conduit avec Jean-René Lecerf la liste d'union RPR-UDF et malgré des dissidences nombreuses à droite, ils font tous deux leur entrée au Sénat.

Adjointe au maire, puis conseillère municipale déléguée de Valenciennes d'octobre 2001 à mars 2004, elle est alors chargée du contrat de ville et du Grand projet de ville. En mars 2004, elle est à nouveau conseillère régionale (tête de liste UDF) et présidente du groupe UDF de la région Nord-Pas-de-Calais ; elle préside également la commission Aménagement du territoire et politique de la ville.

Parmi ses travaux au Sénat, on peut retenir sa réflexion sur les minima sociaux et le retour à l'activité ainsi que sa proposition de loi pour renforcer la parité dans les exécutifs municipaux.

#### Secrétaire d'État de Nicolas Sarkozy et tête de liste aux élections régionales

Le 19 juin 2007, Valérie Létard est nommée secrétaire d'État chargée de la Solidarité dans le gouvernement Fillon II, par le président de la République, Nicolas Sarkozy. Ses premiers axes de travail incluent un Plan Solidarité-Grand âge 2007-2012, l'amélioration de la scolarisation des enfants handicapés, l'intégration des aveugles dans la vie de la cité, la lutte contre les violences faites aux femmes, l'égalité professionnelle hommes-femmes, ainsi qu'un plan pour les métiers de la dépendance. Elle publie fin 2011 un bilan du second Plan Autisme, qui a permis la création de 70 % des places prévues et une nette progression de la scolarisation des enfants et adolescents autistes, avec 25 000 intégrations en milieu scolaire.

Le 9 mars 2008, en seconde position sur la liste de Dominique Riquet, Valérie Létard est réélue conseillère municipale de Valenciennes dès le premier tour des élections municipales. Le 15 mars, elle devient première adjointe lors de l'entrée en fonction du nouveau conseil. Le 10 avril suivant, elle est élue, avec 83 % des suffrages exprimés, présidente de la Communauté d'agglomération de Valenciennes Métropole.

Elle est nommée secrétaire d'État auprès du ministre d'État, ministre de l'Écologie, du Développement durable et de la Mer, chargée des Technologies vertes et des négociations sur le climat, à la faveur d'un remaniement du second gouvernement Fillon, le 23 juin 2009. Le 26 juillet 2009, elle annonce que « 220 000 emplois verts sont à créer d'ici 2012 ». Après avoir pris part au Sommet de Copenhague, œuvré notamment pour la promotion et l'encadrement des métiers de la croissance verte, mis en place, dans le cadre de « Médiaterre », les volontaires en service civique pour sensibiliser des familles modestes aux éco-gestes, lancer un plan de formation pour les travailleurs sociaux aux éco-gestes, lutter contre la précarité énergétique, elle quitte le gouvernement le 13 novembre 2010 après la nomination du troisième gouvernement de François Fillon.

Le 17 novembre 2009, elle est officiellement désignée par l'UMP et son secrétaire général Xavier Bertrand pour conduire la liste d'union UMP-Nouveau centre-Gauche moderne aux élections régionales de 2010 pour la région Nord-Pas-de-Calais. Après une campagne engagée face à Daniel Percheron (PS) et Marine Le Pen, la liste « Ma Vie, Ma Région, Mon Avenir avec Valérie Létard » réalise 19 % au premier tour et 25,91 % au deuxième tour à chaque fois devant la liste de Marine Le Pen. Le groupe Majorité présidentielle se voit attribuer 22 sièges.

En quittant le gouvernement Fillon avec Jean-Louis Borloo, elle annonce vouloir se consacrer à sa Communauté d'agglomération de Valenciennes Métropole, retrouver son siège au Sénat et participer à la création d'une « confédération des centres » après la réunion du 15 novembre 2010 à l'Assemblée nationale. En mai 2011, elle contribue à la création de l'Alliance républicaine, écologiste et sociale.

#### Sénatrice, présidente de Valenciennes Métropole et membre fondatrice de l'UDI

En tant que présidente de Valenciennes Métropole, Valérie Létard a notamment lancé les trois pôles d'excellence du Valenciennois : transports terrestres, numérique et logistique, le lancement du centre d'expositions et des congrès de Valenciennes Métropole, le lancement de l'Institut de recherche supérieur Railenium, le lancement de Transalley - Technopôle Transports Terrestres, le lancement du programme Rîves Créatives de l’Escaut, le lancement du Terminal Dock Seine Nord Europe Escaut, réalise le programme ANRU du Valenciennois, le programme national de requalification des quartiers anciens dégradés, inaugure le stade du Hainaut, concourt au déploiement du très haut débit, lance l'éco-port de plaisance de Valenciennes, lance le complexe hôtelier et immobilier hôpital du Hainaut de Valenciennes, accueille de grands acteurs économiques (CIV, Oxylane, Agrati, etc.), participe à la rénovation des boulevards extérieurs de Valenciennes, à la création de Plateau multisports et Commerce de proximité Communes Rurales, au lancement de l'aéroport Valenciennes-Denain Charles Nungesser, etc.

Le 25 septembre 2011, Valérie Létard est réélue sénatrice du Nord en deuxième position sur la liste « République et territoires, des élus libres et responsables pour l’avenir du Nord », emmenée par Jean-René Lecerf (UMP-NC-NI) non investie par l'UMP et soutenue par Jean-Pierre Raffarin et Jean-Louis Borloo. Le 1er octobre suivant, après avoir été désignée candidate du groupe centriste à la présidence du Sénat, elle obtient 29 voix face à Gérard Larcher et Jean-Pierre Bel. Elle est ainsi la deuxième femme à se présenter à la présidence du Sénat, plusieurs années après la communiste Hélène Luc. Le 8 décembre 2011, elle est élue vice-présidente de l'Association des maires de France, rapporteur de la commission Villes et territoires urbains.

Valérie Létard rend le 12 janvier 2012 son rapport sur l'évaluation de l'impact du Plan Autisme 2008-2010 à Roselyne Bachelot, ministre des Solidarités et de la Cohésion sociale, un document qui fait référence. À la faveur de la démission de Dominique Riquet de la mairie de Valenciennes le 21 juin 2012, elle est, en tant que première adjointe, chargée de l'intérim du 22 au 28 juin, avant l'élection de Laurent Degallaix comme nouveau maire de la ville.

Valérie Létard est membre fondateur de l'Union des démocrates et indépendants. Elle est élue présidente de l'UDI Nord le 11 juin 2013 à Valenciennes lors de la convention départementale. Elle est nommée 2e vice-présidente, chargée de la Rénovation Urbaine et Politique de la Ville, après l'élection de Jean-Christophe Lagarde au poste de président, le 13 novembre 2014.

Le 15 juillet 2013, Valérie Létard et Jean-Louis Touraine, député PS du Rhône, acceptent une mission de concertation en vue de la réforme du droit d'asile confiée par Manuel Valls, ministre de l'Intérieur. Les conclusions de la mission parlementaire sont rendues le 28 novembre 2013 dans le contexte médiatique particulier de l'affaire Leonarda.

Au cours des élections municipales de 2014, sur la liste de Laurent Degallaix, elle est réélue conseillère municipale de Valenciennes le 30 mars, et est réélue présidente de la Communauté d'agglomération de Valenciennes Métropole le 17 avril 2014. À la faveur du renouvellement du Sénat en septembre 2014, elle intègre le bureau du Sénat sous la présidence de Gérard Larcher. Présidente de l’UDI Nord, elle prend une part active aux élections municipales de 2014 et à l’élection de Jean-René Lecerf à la présidence du conseil départemental du Nord.

#### Première vice-présidente du conseil régional des Hauts-de-France

Le 26 septembre 2015, elle est nommée vice-présidente du conseil national des villes. Chef de file UDI Nord-Pas-de-Calais-Picardie et candidate à l'élection régionale de 2015 en Nord-Pas-de-Calais-Picardie, elle est tête de la liste de Xavier Bertrand sur le département du Nord, qui totalise 24,96 % des suffrages au premier tour et 57,77 % au second tour, remportant ainsi l'élection. Elle est ensuite élue première vice-présidente du conseil régional des Hauts-de-France, chargée des solidarités avec les territoires, contrat de plan État-région, des fonds structurels européens, et de l'aménagement du territoire. Le 9 mars 2016, elle est élue Présidente de l'Établissement Public Foncier (EPF) du Nord-Pas-de-Calais. Lors de la Séance Plénière du Conseil Régional des Hauts-de-France du 8 juillet 2016, elle présente et fait adopter les délibérations Cadre Politique Aménagement du Territoire et Ingénierie Territoriale des Hauts-de-France. Le 2 décembre 2016, elle est élue par ses collègues Présidente du Groupe UDI-Union Centriste du Conseil Régional des Hauts-de-France (42 élus).

Elle soutient Alain Juppé pour la primaire présidentielle des Républicains de 2016. Le 2 mars 2017, dans le cadre de l'affaire Fillon, elle demande le retrait du candidat LR François Fillon à l'élection présidentielle. Reçue par le Président Emmanuel Macron le 13 mai 2017, elle refuse d’être ministre du Logement et de la Ville du gouvernement d’Édouard Philippe le 15 mai 2017.

Elle s'implique avec des élus des Hauts-de-France pour la réalisation du canal Seine-Nord-Europe.

#### Vice-présidente du Sénat

Le 24 septembre 2017, Valérie Létard est réélue sénatrice du Nord pour la troisième fois, en tant que tête de la liste « République & Territoires » qui rassemble des élus UDI, LR et non-inscrits, qui obtient le meilleur score dans le département du Nord. Olivier Henno est également élu sur cette liste. Ils intègrent le groupe Union Centriste. Elle est nommée vice-présidente du Sénat le 4 octobre 2017 et secrétaire de la commission des affaires économiques le 5 octobre. Au sein du bureau du Sénat, elle préside la délégation chargée du travail parlementaire, de la législation en commission, des votes et du contrôle.

Le 23 novembre 2017, en raison de l'application de la nouvelle loi sur le cumul des mandats, Valérie Létard cède les délégations de sa vice-présidence au conseil régional des Hauts-de-France à Nicolas Lebas et Salvatore Castiglione, nouveau vice-président, ainsi que la présidence de l'EPF du Nord-Pas-de-Calais. Elle reste conseillère régionale et présidente du groupe UDI-UC, composé de 42 élus. Le 17 mars 2018, à l'occasion du Congrès de l'UDI, Valérie Létard est nommée présidente du conseil national du parti.

En tant que sénatrice du Nord et ancienne présidente de Valenciennes Métropole, elle est impliquée sur le dossier de reprise de l'aciérie Ascoval à Saint-Saulve. Elle initie la mise en place d'une mission d'information sénatoriale sur l'avenir de la sidérurgie, dont elle est nommée rapporteur. Remis le 10 juillet, le rapport formule 7 priorités et 30 propositions, dont celle d'un « ministère de l'Industrie, avec une administration, des ressources budgétaires, avec un État stratège pour être au rendez-vous des enjeux industriels ». Le 13 mars 2019 est adopté sa proposition de loi destinée à améliorer la lisibilité du droit en supprimant des lois obsolètes. Le 12 avril 2019, elle dépose une proposition de loi tendant à protéger les mineurs des usages dangereux du protoxyde d'azote et le 24 juin 2019, elle intègre le Conseil d'Administration de l'ANRU.

Le 11 décembre 2019, sa proposition de loi tendant à protéger les mineurs des usages dangereux du protoxyde d'azote est adoptée à l'unanimité au Sénat. La mesure phare est l’interdiction de la vente aux mineurs, y compris sur les sites de commerce en ligne. Le texte propose aussi de pénaliser l’incitation d’un mineur à faire un usage détourné d’un produit de consommation courante pour en obtenir des effets psychoactifs. Il prévoit encore d’accompagner la politique de prévention menée à l’école. La loi est définitivement adoptée le 25 mai 2021 et promulguée le 1er juin 2021.

Le 6 octobre 2020, Valérie Létard est confirmée dans ses fonctions de vice-présidente de la haute assemblée lors du renouvellement des instances de l'institution.

Candidate aux élections départementales des 20 et 27 juin 2021, Valérie Létard est élue, avec Laurent Degallaix, conseillère départementale sur le canton de Valenciennes. Le Président du Département Christian Poiret lui confie une Mission d'information et d'évaluation de l'allocation personnalisée d'autonomie (APA). Elle s'investit également sur la mise en place d'un RSA d'urgence à destination des victimes de violences conjugales et mène l'expérimentation sur le valenciennois. En mai 2022, son nom est évoqué pour le poste de Première ministre à la suite de la réélection du président Emmanuel Macron. En septembre 2022, elle dépose une proposition de loi créant une aide universelle d'urgence pour les victimes des violences conjugales, un texte issu de l'expérimentation de terrain et qui vise à répondre à la problématique de l'emprise financière du conjoint violent. « Cette aide universelle s'adresse à l'ensemble des victimes de violences conjugales, quels que soient leur sexe, leur origine, leur statut ou leur âge. Elle se veut un moyen de rendre la dignité, de permettre aux victimes de relever la tête, de s'extraire d'une situation personnelle qui n'apporte que souffrance et douleur » explique-t-elle dans l'hémicycle du Sénat. Le texte est adopté à l'unanimité au Sénat le 20 octobre 2022 en première lecture, puis adopté également en première lecture et à l'unanimité à l'Assemblée nationale le 16 janvier 2023. Voté conforme, et à l'unanimité, en deuxième lecture le 16 février 2023 au Sénat, il est définitivement adopté. La loi Létard est promulguée le 28 février 2023. Un décret doit encore venir fixer la date précise d'application de cette aide. Elle sera accessible, au plus tard, d'ici fin 2023, auprès des caisses d'allocations familiales (CAF) et des caisses de la Mutualité sociale agricole (MSA).

Elle annonce en octobre 2022 qu'après trois mandats de sénatrice, elle ne se représentera pas aux élections sénatoriales de 2023. Elle s'implique sur la question du Zéro artificialisation nette, à la suite d'un décret du gouvernement. Au Sénat, elle préside ainsi la commission spéciale relative à la mise en application des mesures de « Zéro artificialisation nette » de la loi Climat-résilience, dépose une proposition de loi le 14 décembre 2022. Elle est adoptée au Sénat le 16 mars 2023, à l'Assemblée nationale le 27 juin 2023 et fait l'objet d'une commission mixte paritaire, qu'elle préside, conclusive le 6 juillet 2023 : « Nous avons su surmonter nos profondes divergences pour élaborer un texte qui réponde concrètement aux difficultés et blocages qui nous sont remontés des élus et des acteurs du ZAN. Nous préservons les capacités de développement des territoires, sans rien sacrifier de nos ambitions écologiques ».

Au Sénat, elle a été l'auteure du rapport sur l'affaire Benalla.

#### Députée de la 21e circonscription du Nord

Le 7 juillet 2024, elle est élue députée de la 21e circonscription du Nord avec 51,6 % des voix au second tour face à la candidate du Rassemblement national. Elle intègre le groupe Libertés, indépendants, outre-mer et territoires (LIOT).

Son nom est un temps évoqué pour la présidence de l'Assemblée nationale. Elle refuse cette idée, affirmant privilégier l'investissement sur son territoire.

#### Ministre du Logement et de la Rénovation urbaine

Dans le gouvernement Michel Barnier, constitué le 21 septembre 2024, elle est nommée ministre du Logement et de la Rénovation urbaine.

En tant que ministre, elle souhaite développer le prêt à taux zéro, construire davantage de logements sociaux (16 000 logements sociaux supplémentaires pour 2027) et simplifier le droit de l’urbanisme. Elle déclare vouloir une baisse du taux du Livret A pour « donner une respiration aux bailleurs sociaux ». Elle veut aussi défiscaliser les donations aux enfants et petits-enfants afin de mobiliser davantage « les capitaux dormants ».

En avril 2025, elle effectue une tournée de deux jours en Martinique afin de « renforcer la collaboration entre l'État et les acteurs locaux, de simplifier les procédures et de mobiliser davantage de ressources, notamment en matière de foncier et d’ingénierie, pour accélérer la production de logements et garantir un accès équitable à tous ».

## Détail des mandats et fonctions

### Mandats et nominations actuels

- Depuis le 21 septembre 2024 : ministre du Logement et de la Rénovation urbaine
- Depuis le 8 juillet 2024 : députée de la 21e circonscription du Nord
- Depuis le 27 juin 2021 : conseillère départementale du Nord
- Depuis le 16 novembre 2021 : présidente de la fédération UDI du Nord

### Anciens mandats et nominations

- 22 mars 1998 - 22 juin 2010 : conseillère régionale du Nord-Pas-de-Calais
- octobre 2001 - mars 2004 : adjointe au maire de Valenciennes, puis conseillère municipale déléguée, chargée de la Politique de la ville
- octobre 2001 - juillet 2007 et décembre 2010 - septembre 2023 : sénatrice du Nord
- 2002 - 2007 : membre du comité exécutif de l'UDF
- octobre 2004 - juillet 2007 : vice-présidente de la commission des Affaires sociales du Sénat
- octobre 2004 - juillet 2007 : vice-présidente du groupe Union centriste – UDF au Sénat
- 2004 - 2010 : présidente du groupe centriste au conseil régional du Nord-Pas-de-Calais
- avril 2006 - mai 2007 : vice-présidente de l'Union pour la démocratie française (UDF)
- 19 juin 2007 - 23 juin 2009 : secrétaire d'État auprès du ministre du Travail, des Relations sociales et de la Solidarité, chargée de la Solidarité dans le gouvernement Fillon II
- 15 mars 2008 - 30 mars 2014 : première adjointe au maire de Valenciennes
- 10 avril 2008 - 12 janvier 2016 : présidente de la communauté d'agglomération Valenciennes Métropole
- mai 2008 - janvier 2013 : présidente déléguée du Nouveau Centre
- 22 juin 2012 - 28 juin 2012 : maire de Valenciennes (par intérim)
- 23 juin 2009 - 13 novembre 2010 : secrétaire d'État auprès du ministre d'État, ministre de l'Écologie, de l'Énergie, du Développement durable et de la Mer, chargée des Technologies vertes et des Négociations sur le climat dans le gouvernement Fillon II
- 8 décembre 2011 - 31 décembre 2015 : vice-présidente de l'Association des maires de France, rapporteur de la commission Villes et territoires urbains
- 10 juin 2013 - 1er octobre 2019 : présidente de la fédération UDI du Nord
- janvier 2013 - mars 2018 : 2e vice-présidente de l'Union des démocrates et indépendants (UDI)
- 30 mars 2014 - 12 janvier 2016 : conseillère municipale de Valenciennes
- septembre 2015 - décembre 2016 : vice-présidente du Conseil national des villes
- 13 décembre 2015 - 23 novembre 2017 : 1re vice-présidente du conseil régional des Hauts-de-France
- 13 décembre 2015 - 1er juillet 2021 : conseillère régionale des Hauts-de-France
- mars 2016 - janvier 2018 : présidente du conseil d'administration de l'Établissement public foncier du Nord-Pas-de-Calais
- 2 décembre 2016 - 1er juillet 2021 : présidente du groupe UDI – Union centriste au conseil régional des Hauts-de-France
- 4 octobre 2017 - 24 septembre 2023 : vice-présidente du Sénat
- 17 mars 2018 - 1er octobre 2019 : présidente du conseil national de l'Union des démocrates et indépendants (UDI)

## Pour approfondir

### Articles connexes

- Femmes ministres en France
- Gouvernement François Fillon (2)
- Gouvernement Michel Barnier
- Gouvernement François Bayrou

### Liens externes

- Site officiel
- Ressources relatives à la vie publique :
  - Assemblée nationale
  - Haute Autorité pour la transparence de la vie publique
  - Sénat
  - Base Sycomore

## Notes et références

### Notes

### Références

- Portail de la politique française
- Portail du Nord-Pas-de-Calais
"
,
    "distance": -13.429688890770539,
  },
]
`;
